overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -0,0 +1,328 @@
+	}
+	if p&apparmor.MayOpenPermission != 0 {
+		perms = append(perms, PermissionOpen)
+	}
+	if p&apparmor.MayRenamePermission != 0 {
+		perms = append(perms, PermissionRename)
+	}
+	if p&apparmor.MaySetAttrPermission != 0 {
+		perms = append(perms, PermissionSetAttr)
+	}
+	if p&apparmor.MayGetAttrPermission != 0 {
+		perms = append(perms, PermissionGetAttr)
+	}
+	if p&apparmor.MaySetCredentialPermission != 0 {
+		perms = append(perms, PermissionSetCred)
+	}
+	if p&apparmor.MayGetCredentialPermission != 0 {
+		perms = append(perms, PermissionGetCred)
+	}
+	if p&apparmor.MayChangeModePermission != 0 {
+		perms = append(perms, PermissionChangeMode)
+	}
+	if p&apparmor.MayChangeOwnerPermission != 0 {
+		perms = append(perms, PermissionChangeOwner)
+	}
+	if p&apparmor.MayChangeGroupPermission != 0 {
+		perms = append(perms, PermissionChangeGroup)
+	}
+	if p&apparmor.MayLockPermission != 0 {
+		perms = append(perms, PermissionLock)
+	}
+	if p&apparmor.MayExecuteMapPermission != 0 {
+		perms = append(perms, PermissionExecuteMap)
+	}
+	if p&apparmor.MayLinkPermission != 0 {
+		perms = append(perms, PermissionLink)
+	}
+	if p&apparmor.MayChangeProfileOnExecPermission != 0 {
+		perms = append(perms, PermissionChangeProfileOnExec)
+	}
+	if p&apparmor.MayChangeProfilePermission != 0 {
+		perms = append(perms, PermissionChangeProfile)
+	}
+	if p&^apparmor.AllFilePermissionMask != 0 {
+		return perms, ErrUnrecognizedFilePermission
+	}
+	return perms, nil
+}
+
+// Returns true if the given permissions list contains the given permission, else false.
+func PermissionsListContains(list []PermissionType, permission PermissionType) bool {
+	for _, perm := range list {
+		if perm == permission {
+			return true
+		}
+	}
+	return false
+}
+
+var allowablePathPatternRegexp = regexp.MustCompile(`^(/|(/[^/*{}]+)*(/\*|(/\*\*)?(/\*\.[^/*{}]+)?)?)$`)
+
+// Checks that the given path pattern is valid.  Returns nil if so, otherwise
+// returns ErrInvalidPathPattern.
+func ValidatePathPattern(pattern string) error {
+	if !allowablePathPatternRegexp.MatchString(pattern) {
+		return ErrInvalidPathPattern
+	}
+	return nil
+}
+
+// Checks that the given outcome is valid.  Returns nil if so, otherwise
+// returns ErrInvalidOutcome.
+func ValidateOutcome(outcome OutcomeType) error {
+	switch outcome {
+	case OutcomeAllow, OutcomeDeny:
+		return nil
+	default:
+		return ErrInvalidOutcome
+	}
+}
+
+// ValidateLifespanParseDuration checks that the given lifespan is valid and
+// that the given duration is valid for that lifespan.  If the lifespan is
+// LifespanTimespan, then duration must be a positive integer representing the
+// number of seconds for which the rule should be valid. Otherwise, it must be
+// 0. Returns an error if any of the above are invalid, otherwise computes the
+// expiration time of the rule based on the current time and the given duration
+// and returns it.
+func ValidateLifespanParseDuration(lifespan LifespanType, duration int) (string, error) {
```

Comment:
Should the expiration time be returned as a `time.Time` instead of a string? Probably not. More likely, a `time.Time` would be used within the stored rules, and converted to a string whenever a rule is outputted -- this would require separate types for internal and external access rules, which is undesirable, but would reduce the need to convert from rule expiration string to a `time.Time` whenever matching a path against stored rules. Notably, this only needs to occur for rules which have a lifespan type of `LifespanTimespan`, which is likely not many rules. So continuing to store and output expiration times as strings is likely fine.